### PR TITLE
TagBox\SelectBox: A popup window is displayed with a gap when it's opened the second time if floating labels are used (T1104913)

### PR DESCRIPTION
**Pull request by EugeniyKiyashko**

### CODE REVIEW

**Review by ksercs — DISMISSED** [review]

File: js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 1]
```diff
@@ -617,7 +617,30 @@ const DropDownEditor = TextBox.inherit({
     },
 
     _popupPositionedHandler: function(e) {
-        e.position && this._popup.$overlayContent().toggleClass(DROP_DOWN_EDITOR_OVERLAY_FLIPPED, e.position.v.flip);
+        const $popupOverlayContent = this._popup.$overlayContent();
+        const isOverlayFlipped = e.position.v.flip;
+        const labelMode = this.option('labelMode');
```
Comment (EugeniyKiyashko) [inline js/ui/drop_down_editor/ui.drop_down_editor.js]: Let's use destructuring assignment to get these values from option's object

File: js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 2]
```diff
@@ -617,7 +617,30 @@ const DropDownEditor = TextBox.inherit({
...
+        const stylingMode = this.option('stylingMode');
+
+        const isNeedMoved = (labelMode, stylingMode) => (
```
Comment (EugeniyKiyashko) [inline js/ui/drop_down_editor/ui.drop_down_editor.js]: Why you decided that we need to do anything for underlined mode?

File: js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 3]
```diff
@@ -617,7 +617,30 @@ const DropDownEditor = TextBox.inherit({
...
+            (labelMode === 'static' && stylingMode === 'underlined') ||
+            (labelMode === 'floating' && stylingMode === 'underlined')) &&
+            isOverlayFlipped && this._label.isVisible()
+        );
+
+        if(e.position) {
+            $popupOverlayContent.toggleClass(DROP_DOWN_EDITOR_OVERLAY_FLIPPED, isOverlayFlipped);
+        }
+
+        if(isNeedMoved(labelMode, stylingMode)) {
```
Comment (EugeniyKiyashko) [inline js/ui/drop_down_editor/ui.drop_down_editor.js]: Declare the function inside other function it's not good idea. Will better if it be a method of our component that contained correspond logic.

File: playground/jquery.html [hunk 4]
```diff
@@ -42,4 +42,4 @@
         });
     </script>
 </body>
-</html>
```
Comment (EugeniyKiyashko) [inline playground/jquery.html]: Lets revert this change

File: testing/testcafe/tests/editors/tagBox/label.ts [hunk 5]
```diff
@@ -39,5 +44,53 @@ themes.forEach((theme) => {
         multiline: true,
       }, true, '#otherContainer');
     });
+
+    labelMods.forEach((labelMode) => {
+      test(`Label shouldn't be cutted for dxTagBox ${theme} in stylingMode=${stylingMode}, labelMode=${labelMode} (T1104913)`, async (t) => {
+        const tagBox = new TagBox('#container');
+
+        await t.click(tagBox.element);
+        await t.wait(500);
```
Comment (EugeniyKiyashko) [inline testing/testcafe/tests/editors/tagBox/label.ts]: This timeout required for this case?

File: js/ui/drop_down_editor/ui.drop_down_editor.js (hunk 2 above)
Comment (marker-dao) [reply js/ui/drop_down_editor/ui.drop_down_editor.js]: https://codepen.io/marker-dao/pen/WNzVVZG

**Review by marker-dao — DISMISSED** [review]

File: testing/testcafe/tests/editors/tagBox/label.ts [hunk 6]
```diff
@@ -2,8 +2,11 @@ import { compareScreenshot } from 'devextreme-screenshot-comparer';
 import { changeTheme } from '../../../helpers/changeTheme';
 import url from '../../../helpers/getPageUrl';
 import createWidget from '../../../helpers/createWidget';
+import { setAttribute } from '../../navigation/helpers/domUtils';
+import TagBox from '../../../model/tagBox';
 
 const stylingMods = ['outlined', 'underlined', 'filled'];
+const labelMods = ['static', 'floating', 'hidden'];
```
Comment (ksercs) [inline testing/testcafe/tests/editors/tagBox/label.ts]:
```suggestion
const labelModes = ['static', 'floating', 'hidden'];
```

File: testing/testcafe/tests/editors/tagBox/label.ts [hunk 7]
```diff
@@ -39,5 +44,50 @@ themes.forEach((theme) => {
...
     });
+
+    labelMods.forEach((labelMode) => {
+      test(`Label shouldn't be cutted for dxTagBox ${theme} in stylingMode=${stylingMode}, labelMode=${labelMode} (T1104913)`, async (t) => {
+        const tagBox = new TagBox('#container');
+
+        await t.click(tagBox.element);
+
+        await t
+          .expect(await compareScreenshot(t, `label-tag-box-styleMode=${stylingMode},labelMode=${labelMode},theme=${theme.replace(/\./g, '-')}.png`))
```
Comment (ksercs) [inline testing/testcafe/tests/editors/tagBox/label.ts]: Lee's move etalon name to a variable to reduce copypaste (55 and 62 row)?

File: js/ui/drop_down_editor/ui.drop_down_editor.js [hunk 8]
```diff
@@ -617,7 +617,23 @@ const DropDownEditor = TextBox.inherit({
...
+
+        const $popupOverlayContent = this._popup.$overlayContent();
+        const isOverlayFlipped = e.position.v.flip;
+        const shouldIndentForLabel = labelMode !== 'hidden' && stylingMode === 'outlined';
+
+        if(e.position) {
+            $popupOverlayContent.toggleClass(DROP_DOWN_EDITOR_OVERLAY_FLIPPED, isOverlayFlipped);
+        }
+
+        if(isOverlayFlipped && shouldIndentForLabel && this._label.isVisible()) {
```
Comment (ksercs) [inline js/ui/drop_down_editor/ui.drop_down_editor.js:630]: Hm, i think `this._label.isVisible()` is always `true` if `labelMode !== 'hidden'`, not?

File: testing/testcafe/tests/editors/tagBox/label.ts (hunk 7 above)
Comment (marker-dao) [reply testing/testcafe/tests/editors/tagBox/label.ts]: [commit](https://github.com/DevExpress/DevExtreme/pull/22471/commits/eb90394b50ac98160b9f9b8a15c21453b70813d6)

File: testing/testcafe/tests/editors/tagBox/label.ts (hunk 6 above)
Comment (marker-dao) [reply testing/testcafe/tests/editors/tagBox/label.ts]: [commit](https://github.com/DevExpress/DevExtreme/pull/22471/commits/eb90394b50ac98160b9f9b8a15c21453b70813d6)

File: js/ui/drop_down_editor/ui.drop_down_editor.js (hunk 8 above)
Comment (marker-dao) [reply js/ui/drop_down_editor/ui.drop_down_editor.js:630]: [commit](https://github.com/DevExpress/DevExtreme/pull/22471/commits/eb90394b50ac98160b9f9b8a15c21453b70813d6)

**Review by marker-dao — DISMISSED** [review]

**Review by ksercs — DISMISSED** [review]

**Review by marker-dao — APPROVED** [review]
